src/ChainRouterGenerateBcLayer.php

```diff
@@ -0,0 +1,160 @@
+                    continue;
+                }
+
+                try {
+                    return $router->generate($name, $parameters, $absolute);
+                } catch (RouteNotFoundException $e) {
+                    $hint = $this->getErrorMessage($name, $router, $parameters);
+                    $debug[] = $hint;
+                    if ($this->logger) {
+                        $this->logger->debug('Router '.get_class($router)." was unable to generate route. Reason: '$hint': ".$e->getMessage());
+                    }
+                }
+            }
+
+            if ($debug) {
+                $debug = array_unique($debug);
+                $info = implode(', ', $debug);
+            } else {
+                $info = $this->getErrorMessage($name);
+            }
+
+            throw new RouteNotFoundException(sprintf('None of the chained routers were able to generate route: %s', $info));
+        }
+
+        private function getErrorMessage($name, $router = null, $parameters = null)
+        {
+            if ($router instanceof VersatileGeneratorInterface) {
+                // the $parameters are not forced to be array, but versatile generator does typehint it
+                if (!is_array($parameters)) {
+                    $parameters = [];
+                }
+                $displayName = $router->getRouteDebugMessage($name, $parameters);
+            } elseif (is_object($name)) {
+                $displayName = method_exists($name, '__toString')
+                    ? (string) $name
+                    : get_class($name)
+                ;
+            } else {
+                $displayName = (string) $name;
+            }
+
+            return "Route '$displayName' not found";
+        }
+    }
+} else {
+    /**
+     * @internal
+     */
+    class ChainRouterGenerateBcLayer
+    {
+        /**
+         * Loops through all registered routers and returns a router if one is found.
+         * It will always return the first route generated.
+         */
+        public function generate($name, $parameters = [], $absolute = UrlGeneratorInterface::ABSOLUTE_PATH)
+        {
+            $debug = [];
+
+            if ($name && !is_string($name)) {
+                @trigger_error('Passing an object as the route name is deprecated in symfony-cmf/Routing v2.2 and will not work in Symfony 5.0. Pass an empty route name and the object as "_cmf_route" parameter in the parameters array.', E_USER_DEPRECATED);
+            }
+
+            foreach ($this->all() as $router) {
+                // if $router does not announce it is capable of handling
+                // non-string routes and $name is not a string, continue
+                if ((($name && !is_string($name)) || (array_key_exists('_cmf_route', $parameters) && is_object($parameters['_cmf_route']))) && !$router instanceof VersatileGeneratorInterface) {
+                    continue;
+                }
+
```

Comment:
for BC reasons, we should replace empty string $name with the value of _cmf_route here. otherwise the supports call only works with the deprecated mode, and old generators won't work.

Reply:
I've added a different bc layer on line 52, so we already use the new way. Old code will still work as the generators will still check the name for being an object